### ISSUE TITLE
hello caffe 예제 컴파일 시에 caffe.hpp include 문제

### ISSUE THREAD

**Issue by garion9013** [koosyong/caffestudy#17]
안녕하세요^^
OSX에서 koosyong님께서 올려주신 예제를 천천히 따라해보고 있는데요.
제가 설치한 환경(wiki에 적어놓은...)에서 cmake관련 설정을 잘못한 것인지
camke관련 설정 후에 make를 하면, caffe.hpp 파일을 찾지 못하네요.
강제로 -I 옵션을 집어넣거나 절대경로로 하면 일단 해결될것 같긴한데,
혹시 제가 잘못 설치한 부분이 있나 궁금해서요.
답변 부탁드립니다ㅎㅎ

**Comment by garion9013** [comment]
혹시 caffe를 빌드할 때, 애초에 cmake로 해야 예제처럼 cmake 관련 설정이 자동으로 되는 건가요?

**Comment by super525252** [comment]
오! 저도 이부분에서 같은 에러가 계속 나서 궁금하던  차였는데

**Comment by garion9013** [comment]
혹시 osx 쓰시는건가요?
예제는 아마 빌드할때 cmake 이용해서 빌드한 경우를 가정하고 만드신것 같아요ㅎㅎ

**Comment by kanggeon** [comment]
error가 나는 이유는 CMakeLists에 find_package(Caffe) 부분 때문인데, 말씀하신 것처럼 Caffe 빌드를 CMake로 하셔야 해당 커멘드가 지원이 됩니다. 
https://github.com/BVLC/caffe/pull/1667

**Comment by Jacob-Kang** [comment]
혹시 "Caffe 빌드를 CMake로 하셔야 해당 커멘드가 지원됩니다."
이게 무슨 말씀이신지 좀더 자세히 설명해주실 수 있으신가요?

caffe 빌드 할 때 Makefile.config 수정해서 make 로 했는데
이걸 cmake 로 해야 한다는 말씀이신가요?
지금 "ex1_hellocaffe" 를 해볼려고하는데, CMakeLists 는 처음해보는 거라 find_package(Caffe)  이부분에서 막히네요...
단순히 find_package(Caffe) 라고 해주면 안되는건가요??

**Comment by jslee02** [comment]
Caffe 빌드하고 설치를 하지 않아서 생기는 문제인듯합니다. 설치는 빌드하고난뒤 그 디렉토리에서
```
$ make install
```
하시면 될거에요.

그리고 예제는 cmake를 이용해서 하셔야하는데, `CMakeLists.txt`를 수정할 필요없이 바로 다음 커맨드로 될거에요.
```
$ cd ex1_hellocaffe
$ mkdir build
$ cd build
$ cmake ..
$ make
```

**Comment by Jacob-Kang** [comment]
감사합니다 해결했습니다!
말씀처럼 Caffe 를 cmake 하고 make install 를 하지 않아서 그랬던 것 같습니다.

지금 간단하게 네트워크 학습을 시켜보고싶은데 마땅한 예제가 없네요.. 혹시 추천해주실만한 예제가 있을까요?

우선 ex1돌려봤고,
caffe 가 제공해주는 cpp_classification 를 돌려봤는데 이건 단순히 이미 만들어져있는 네트워크를 불러와서 테스트해보는 소스이더군요..
혹시 직접 학습시키는 간단한 예제가 없을까요..?

**Comment by jslee02** [comment]
해결하셨다니 다행이네요! 그런데 예제는 제가 잘 모르겠네요.. 아마 다음분이 알려주실거에요. :sweat:

**Comment by Jacob-Kang** [comment]
혹시 기초적인 질문들을 할려면 어디다가 올려야할까요..? 마땅히 자유로운 질문을 올릴곳이 없네요.
caffe 가 사용하는 네트워크는 convolution neural network라고 알고 있습니다.
제가 이제 막 시작한 초보라서, 제가 아는 딥러닝은 deep neural network 를 학습시키는 것이고, backpropagation 의 학습시간의 단점을 극복시킨 dropout 과 RBM 으로 인해 최근들어 다시 떠올르고 있다고 들었습니다.
근데 딥러닝 프레임워크라 불리는 caffe 는 제가 알고 있는 내용과 어떤 연관성이 있는지 잘 모르겠어요.

deep nenural network 와 convolution neural network 는 서로 다른 네트워크인지.
caffe 에서 이 네트워크를 학습시키기 위해 사용하는 알고리즘은 무엇인지. 이런것들에 대해서 알고싶어요

**Comment by goddoe** [comment]
머신 러닝이나 뉴럴 넷을 인터넷에서 쉽게 공부하기 좋은 곳인 이 사이트인것같더라구요
http://deepcumen.com

그리고
http://cs231n.github.io
여긴 스탠포드에서 강의했던 CNN 수업인데 들으시면 도움 되실거에요 ㅎㅎ

Deep Neural Network 는 어떤 뉴럴넷이던간에 깊은 층의 레이어와 수많은 노드들을 이용하고 많은 수의 데이터들을 이용해서 학습하면 딥 뉴럴 넷입니다. 

이전에는 노드 숫자도 몇십개, 레이어 층도 2~3 층 정도 였는데  이젠 GPGPU를 이용한 병렬처리와 많은 수의 학습 데이터들로 여러층에 걸쳐진 노드들의 웨이트들을 학습시킬 수 있고 오버피팅 문제들도 해결할 수 있는 많은 방법들이 나왔구요(ex dropout)

CNN도 노드 수를 적게하고 층을 적게하면 그냥 CNN인데 빅데이터를 활용하고 노드수와 레이어 수를 늘리면 Deep 한 CNN인거죠 ㅎㅎ

**Comment by garion9013** [comment]
CNN/DNN 구현 관련 논문에서 주워들은 지식인데,
보통 NN들은 다수의(Convolution, Pooling, LRN, Classifier 등 대략 4-5개 정도) Layer로 구성되어 있다고 합니다.
특히, DNN과 CNN의 가장 큰 차이는 Convolutional layer의 계산을 구성하는 filter가
shared이냐(CNN) fully-connected(DNN)냐에 따라서 구분된다고 들었습니다.
쉽게 말해서 DNN은 CNN보다 neuron끼리 연결된 부분이 많아서 연산량 및 파라미터 양이 더 많다고 할 수 있을것 같습니다.

**Comment by kamchol7021** [comment]
안녕하세요 지금 카페는 설치해서 runtest를 진행했는데 

제가 CMakelist작성을 잘몰라서 질문드립니다..


cmake_minimum_required(VERSION 2.8.8)

project (hellocaffe)

find_package(Caffe)
include_directories(${/home/내경로 /caffe/include})
add_definitions(${Caffe_DEFINITIONS})

add_executable(hellocaffe main.cpp)
target_link_libraries(hellocaffe ${/home/내경로 /caffe/build/lib})

이렇게 하였는데 일단 Caffe 패키지를 찾을 수 없다고 나옵니다 

이것 저것 시도 해봤는데 잘 되지 않네요 혹시 도움 주실 수 있으면 감사하겠습니다
